client/client.go

```diff
@@ -137,3 +145,19 @@ func FromContext(ctx context.Context) Info {
 	}
 	return c
 }
+
+// NewMetadata creates a new Metadata object to use in Info. md is used as-is.
+func NewMetadata(md map[string][]string) Metadata {
+	return Metadata{
+		data: md,
+	}
+}
+
+// Get gets the value of the key from metadata, returning a copy.
+func (m Metadata) Get(key string) []string {
+	vals := m.data[key]
+	ret := make([]string, len(vals))
+	copy(ret, vals)
+
+	return ret
```

Comment:
I think we should check `if len(vals) == 0 return nil`

Reply:
works as-is per test, what's the difference here?

Reply:
You can write your own benchmarks as well:

```
cpu: Intel(R) Core(TM) i9-9880H CPU @ 2.30GHz
BenchmarkGetIf
BenchmarkGetIf-16          	298160968	         3.912 ns/op	       0 B/op	       0 allocs/op
BenchmarkGetIfExists
BenchmarkGetIfExists-16    	40360434	        27.66 ns/op	      16 B/op	       1 allocs/op
BenchmarkGet
BenchmarkGet-16            	128395258	         9.024 ns/op	       0 B/op	       0 allocs/op
BenchmarkGetExists
BenchmarkGetExists-16      	41000848	        28.28 ns/op	      16 B/op	       1 allocs/op

var data = map[string][]string{
	"key": {"value"},
}

func get(key string) []string {
	vals := data[key]
	ret := make([]string, len(vals))
	copy(ret, vals)

	return ret
}

func getIf(key string) []string {
	vals := data[key]
	if len(vals) == 0 {
		return nil
	}
	ret := make([]string, len(vals))
	copy(ret, vals)

	return ret
}

func BenchmarkGetIf(b *testing.B) {
	b.ResetTimer()
	b.ReportAllocs()
	for n := 0; n < b.N; n++ {
		if len(getIf("test")) != 0 {
			b.Fail()
		}
	}
}

func BenchmarkGetIfExists(b *testing.B) {
	b.ResetTimer()
	b.ReportAllocs()
	for n := 0; n < b.N; n++ {
		if len(getIf("key")) == 0 {
			b.Fail()
		}
	}
}

func BenchmarkGet(b *testing.B) {
	b.ResetTimer()
	b.ReportAllocs()
	for n := 0; n < b.N; n++ {
		if len(get("test")) != 0 {
			b.Fail()
		}
	}
}

func BenchmarkGetExists(b *testing.B) {
	b.ResetTimer()
	b.ReportAllocs()
	for n := 0; n < b.N; n++ {
		if len(get("key")) == 0 {
			b.Fail()
		}
	}
}
```

Reply:
Added, thanks